File: packages/ts/tests/integration/binding/from-config.test.ts
// Config-driven construction: `new DirSQL(configPath)`.
//
// Config-defined tables produce one row per matched file. Each row's columns
// come from filesystem facts: the stat virtuals (`path`, `basename`, `dir`,
// `ext`, `size`, `mtime`, `ctime`). Content interpretation is intentionally
// out of scope; for that, register a programmatic Table with your own onFile
// function.

import { mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

// Renamed from `writeFile` to avoid shadowing the `node:fs/promises` import;
// it ensures the parent directory exists before writing.
async function seedFile(path: string, content: string): Promise<void> {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, content);
}

// Each fixture emits its own filesystem-fact columns via an `on-file` hook
// rather than relying on the core injecting them. Hook output overrides
// injection, so results are identical while injection remains and stay green
// once it is removed. `{path}` is the absolute path, `{root}` the index root;
// the relative path is `{path}` with the `{root}/` prefix stripped.
const pathHook = `on-file = '''sh -c 'rel=\${1#"$2"/}; printf "[{\\"path\\":\\"%s\\"}]" "$rel"' sh {path} {root}'''`;
const pathBasenameHook = `on-file = '''sh -c 'rel=\${1#"$2"/}; base=\${1##*/}; printf "[{\\"path\\":\\"%s\\",\\"basename\\":\\"%s\\"}]" "$rel" "$base"' sh {path} {root}'''`;
const basenameHook = `on-file = '''sh -c 'printf "[{\\"basename\\":\\"%s\\"}]" "\${1##*/}"' sh {path}'''`;
const statHook = `on-file = '''sh -c 'rel=\${1#"$2"/}; base=\${1##*/}; case "$rel" in */*) dir=\${rel%/*};; *) dir="";; esac; ext=\${base##*.}; [ "$ext" = "$base" ] && ext=""; size=$(wc -c < "$1" | tr -d " "); mtime=$(stat -c %Y "$1"); printf "[{\\"path\\":\\"%s\\",\\"basename\\":\\"%s\\",\\"dir\\":\\"%s\\",\\"ext\\":\\"%s\\",\\"size\\":%s,\\"mtime\\":%s}]" "$rel" "$base" "$dir" "$ext" "$size" "$mtime"' sh {path} {root}'''`;

describe("new DirSQL(configPath)", () => {
  let dir: string;
  let configPath: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-fromconfig-"));
    configPath = join(dir, ".dirsql.toml");
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  it("produces one row per matched file with stat virtuals", async () => {
    await seedFile(join(dir, "items", "a.csv"), "anything");
    await seedFile(join(dir, "items", "b.csv"), "anything");
    await seedFile(
      configPath,
      `
[[table]]
ddl = "CREATE TABLE items (path TEXT, basename TEXT)"
glob = "items/*.csv"
${pathBasenameHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await db.ready;
    const rows = await db.query(
      "SELECT path, basename FROM items ORDER BY path",
    );
    expect(rows).toHaveLength(2);
    expect(rows[0].path).toBe("items/a.csv");
    expect(rows[0].basename).toBe("a.csv");
    expect(rows[1].path).toBe("items/b.csv");
    expect(rows[1].basename).toBe("b.csv");
  });

  it("rejects a glob placeholder that collides with a declared column", async () => {
    await seedFile(join(dir, "comments", "thread-1", "a.txt"), "x");
    await seedFile(
      configPath,
      `
[[table]]
ddl = "CREATE TABLE comments (thread_id TEXT, basename TEXT)"
glob = "comments/{thread_id}/*.txt"
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await expect(db.ready).rejects.toThrow(/thread_id/);
  });

  it("treats a non-colliding placeholder as a wildcard", async () => {
    await seedFile(join(dir, "comments", "thread-1", "a.txt"), "x");
    await seedFile(join(dir, "comments", "thread-2", "b.txt"), "x");
    await seedFile(
      configPath,
      `
[[table]]
ddl = "CREATE TABLE comments (path TEXT, basename TEXT)"
glob = "comments/{thread_id}/*.txt"
${pathBasenameHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await db.ready;
    const rows = await db.query(
      "SELECT basename FROM comments ORDER BY basename",
    );
    expect(rows).toHaveLength(2);
    expect(rows[0].basename).toBe("a.txt");
    expect(rows[1].basename).toBe("b.txt");
    expect(rows[0].thread_id).toBeUndefined();
  });

  it("exposes the full set of stat virtuals when declared in DDL", async () => {
    const body = "# title\nhello world\n";
    await seedFile(join(dir, "docs", "readme.md"), body);
    await seedFile(
      configPath,
      `
[[table]]
ddl = "CREATE TABLE files (path TEXT, basename TEXT, dir TEXT, ext TEXT, size INTEGER, mtime INTEGER)"
glob = "docs/*.md"
${statHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await db.ready;
    const rows = await db.query(
      "SELECT path, basename, dir, ext, size, mtime FROM files",
    );
    expect(rows).toHaveLength(1);
    const r = rows[0];
    expect(r.path).toBe("docs/readme.md");
    expect(r.basename).toBe("readme.md");
    expect(r.dir).toBe("docs");
    expect(r.ext).toBe("md");
    expect(r.size).toBe(body.length);
    expect(typeof r.mtime).toBe("number");
    expect(r.mtime as number).toBeGreaterThan(0);
  });

  it("respects ignore patterns from config", async () => {
    await seedFile(join(dir, "data", "good.json"), "{}");
    await seedFile(join(dir, "data", "node_modules", "bad.json"), "{}");
    await seedFile(
      configPath,
      `
[dirsql]
ignore = ["**/node_modules/**"]

[[table]]
ddl = "CREATE TABLE items (path TEXT)"
glob = "data/**/*.json"
${pathHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await db.ready;
    const rows = await db.query("SELECT path FROM items");
    expect(rows).toHaveLength(1);
    expect(rows[0].path).toBe("data/good.json");
  });

  it("loads multiple tables", async () => {
    await seedFile(join(dir, "posts", "hello.txt"), "x");
    await seedFile(join(dir, "authors", "alice.txt"), "x");
    await seedFile(
      configPath,
      `
[[table]]
ddl = "CREATE TABLE posts (basename TEXT)"
glob = "posts/*.txt"
${basenameHook}

[[table]]
ddl = "CREATE TABLE authors (basename TEXT)"
glob = "authors/*.txt"
${basenameHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await db.ready;
    const posts = await db.query("SELECT basename FROM posts");
    const authors = await db.query("SELECT basename FROM authors");
    expect(posts).toHaveLength(1);
    expect(authors).toHaveLength(1);
    expect(posts[0].basename).toBe("hello.txt");
    expect(authors[0].basename).toBe("alice.txt");
  });

  it("rejects missing config files", async () => {
    const missing = join(dir, "nonexistent.toml");
    const db = new DirSQL(missing);
    await expect(db.ready).rejects.toThrow();
  });

  it("rejects invalid TOML", async () => {
    await seedFile(configPath, "this is not valid [[[");
    const db = new DirSQL(configPath);
    await expect(db.ready).rejects.toThrow();
  });

  it("rejects table entries missing ddl", async () => {
    await seedFile(
      configPath,
      `
[[table]]
glob = "*.json"
`,
    );
    const db = new DirSQL(configPath);
    await expect(db.ready).rejects.toThrow();
  });
});

File: packages/python/tests/integration/binding/from_config_test.py
"""Binding-tier tests (real core, real fs) for DirSQL(config=).

Config-defined tables produce one row per matched file. Each row's columns
come from the table's `on-file` hook: a small `sh` command that derives the
stat facts (`path`, `basename`, `dir`, `ext`, `size`, `mtime`) from the file
and emits them as a JSON row. Content interpretation beyond that is out of
scope; for richer parsing, register a programmatic Table with your own on_file
function.
"""

import os
import tempfile

import pytest

from dirsql import DirSQL

# `on-file` hooks that emit the stat facts a row needs, derived from the
# file path (`{path}`) relative to the scan root (`{root}`). Emitting a
# superset of the DDL's columns is safe -- undeclared keys are dropped.
_HOOK_PATH = r"""on-file = '''sh -c 'rel=${1#"$2"/}; printf "[{\"path\":\"%s\"}]" "$rel"' sh {path} {root}'''"""
_HOOK_PATH_BASENAME = r"""on-file = '''sh -c 'rel=${1#"$2"/}; base=${1##*/}; printf "[{\"path\":\"%s\",\"basename\":\"%s\"}]" "$rel" "$base"' sh {path} {root}'''"""
_HOOK_STAT = r"""on-file = '''sh -c 'rel=${1#"$2"/}; base=${1##*/}; case "$rel" in */*) dir=${rel%/*};; *) dir="";; esac; ext=${base##*.}; [ "$ext" = "$base" ] && ext=""; size=$(wc -c < "$1" | tr -d " "); mtime=$(stat -c %Y "$1"); printf "[{\"path\":\"%s\",\"basename\":\"%s\",\"dir\":\"%s\",\"ext\":\"%s\",\"size\":%s,\"mtime\":%s}]" "$rel" "$base" "$dir" "$ext" "$size" "$mtime"' sh {path} {root}'''"""


@pytest.fixture
def config_dir():
    with tempfile.TemporaryDirectory() as d:
        yield d


def _write(path, content):
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        f.write(content)


def describe_DirSQL_from_config():
    def describe_basic():
        @pytest.mark.asyncio
        async def it_produces_one_row_per_matched_file(config_dir):
            _write(os.path.join(config_dir, "items", "a.csv"), "anything")
            _write(os.path.join(config_dir, "items", "b.csv"), "anything")
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "items"
ddl = "CREATE TABLE items (path TEXT, basename TEXT)"
glob = "items/*.csv"
"""
                + _HOOK_PATH_BASENAME
                + "\n",
            )

            db = DirSQL(
                root=config_dir, config=os.path.join(config_dir, ".dirsql.toml")
            )
            await db.ready()
            results = await db.query("SELECT path, basename FROM items ORDER BY path")
            assert len(results) == 2
            assert results[0]["path"] == "items/a.csv"
            assert results[0]["basename"] == "a.csv"
            assert results[1]["path"] == "items/b.csv"
            assert results[1]["basename"] == "b.csv"

    def describe_glob_placeholders():
        @pytest.mark.asyncio
        async def it_errors_when_a_placeholder_collides_with_a_column(config_dir):
            _write(
                os.path.join(config_dir, "comments", "thread-1", "a.txt"),
                "hello",
            )
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "comments"
ddl = "CREATE TABLE comments (thread_id TEXT, basename TEXT)"
glob = "comments/{thread_id}/*.txt"
"""
                + _HOOK_PATH_BASENAME
                + "\n",
            )

            db = DirSQL(
                root=config_dir, config=os.path.join(config_dir, ".dirsql.toml")
            )
            with pytest.raises(Exception, match="thread_id"):
                await db.ready()

        @pytest.mark.asyncio
        async def it_treats_a_non_colliding_placeholder_as_a_wildcard(config_dir):
            _write(
                os.path.join(config_dir, "comments", "thread-1", "a.txt"),
                "hello",
            )
            _write(
                os.path.join(config_dir, "comments", "thread-2", "b.txt"),
                "world",
            )
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "comments"
ddl = "CREATE TABLE comments (path TEXT, basename TEXT)"
glob = "comments/{thread_id}/*.txt"
"""
                + _HOOK_PATH_BASENAME
                + "\n",
            )

            db = DirSQL(
                root=config_dir, config=os.path.join(config_dir, ".dirsql.toml")
            )
            await db.ready()
            results = await db.query("SELECT basename FROM comments ORDER BY basename")
            assert len(results) == 2
            assert results[0]["basename"] == "a.txt"
            assert results[1]["basename"] == "b.txt"
            assert "thread_id" not in results[0]

    def describe_stat_virtuals():
        @pytest.mark.asyncio
        async def it_exposes_stat_virtuals(config_dir):
            body = "# title\nhello world\n"
            _write(os.path.join(config_dir, "docs", "readme.md"), body)
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "files"
ddl = "CREATE TABLE files (path TEXT, basename TEXT, dir TEXT, ext TEXT, size INTEGER, mtime INTEGER)"
glob = "docs/*.md"
"""
                + _HOOK_STAT
                + "\n",
            )

            db = DirSQL(
                root=config_dir, config=os.path.join(config_dir, ".dirsql.toml")
            )
            await db.ready()
            results = await db.query(
                "SELECT path, basename, dir, ext, size, mtime FROM files"
            )
            assert len(results) == 1
            r = results[0]
            assert r["path"] == "docs/readme.md"
            assert r["basename"] == "readme.md"
            assert r["dir"] == "docs"
            assert r["ext"] == "md"
            assert r["size"] == len(body)
            assert isinstance(r["mtime"], int)
            assert r["mtime"] > 0

    def describe_ignore():
        @pytest.mark.asyncio
        async def it_respects_ignore_patterns(config_dir):
            _write(os.path.join(config_dir, "data", "good.json"), "{}")
            _write(
                os.path.join(config_dir, "data", "node_modules", "bad.json"),
                "{}",
            )
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[dirsql]
ignore = ["**/node_modules/**"]

[[table]]
name = "items"
ddl = "CREATE TABLE items (path TEXT)"
glob = "data/**/*.json"
"""
                + _HOOK_PATH
                + "\n",
            )

            db = DirSQL(
                root=config_dir, config=os.path.join(config_dir, ".dirsql.toml")
            )
            await db.ready()
            results = await db.query("SELECT path FROM items")
            assert len(results) == 1
            assert results[0]["path"] == "data/good.json"

    def describe_multiple_tables():
        @pytest.mark.asyncio
        async def it_loads_multiple_tables(config_dir):
            _write(os.path.join(config_dir, "posts", "hello.txt"), "x")
            _write(os.path.join(config_dir, "authors", "alice.txt"), "x")
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "posts"
ddl = "CREATE TABLE posts (basename TEXT)"
glob = "posts/*.txt"
"""
                + _HOOK_PATH_BASENAME
                + """

[[table]]
name = "authors"
ddl = "CREATE TABLE authors (basename TEXT)"
glob = "authors/*.txt"
"""
                + _HOOK_PATH_BASENAME
                + "\n",
            )

            db = DirSQL(
                root=config_dir, config=os.path.join(config_dir, ".dirsql.toml")
            )
            await db.ready()
            posts = await db.query("SELECT basename FROM posts")
            authors = await db.query("SELECT basename FROM authors")
            assert len(posts) == 1
            assert len(authors) == 1
            assert posts[0]["basename"] == "hello.txt"
            assert authors[0]["basename"] == "alice.txt"

    def describe_error_handling():
        @pytest.mark.asyncio
        async def it_raises_on_missing_config_file(config_dir):
            db = DirSQL(config=os.path.join(config_dir, "nonexistent.toml"))
            with pytest.raises(Exception):
                await db.ready()

        @pytest.mark.asyncio
        async def it_raises_on_invalid_toml(config_dir):
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                "this is not valid [[[",
            )
            db = DirSQL(config=os.path.join(config_dir, ".dirsql.toml"))
            with pytest.raises(Exception):
                await db.ready()

        @pytest.mark.asyncio
        async def it_raises_on_a_hookless_table(config_dir):
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "files"
ddl = "CREATE TABLE files (path TEXT, size INTEGER)"
glob = "**/*.md"
""",
            )
            db = DirSQL(config=os.path.join(config_dir, ".dirsql.toml"))
            with pytest.raises(Exception, match="FROM './'"):
                await db.ready()

        @pytest.mark.asyncio
        async def it_raises_on_missing_ddl(config_dir):
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "t"
glob = "*.json"
""",
            )
            db = DirSQL(config=os.path.join(config_dir, ".dirsql.toml"))
            with pytest.raises(Exception):
                await db.ready()

    def describe_query_after_config():
        @pytest.mark.asyncio
        async def it_supports_sql_queries_after_config_init(config_dir):
            _write(os.path.join(config_dir, "items", "apple.json"), "x")
            _write(
                os.path.join(config_dir, ".dirsql.toml"),
                """\
[[table]]
name = "items"
ddl = "CREATE TABLE items (basename TEXT, size INTEGER)"
glob = "items/*.json"
"""
                + _HOOK_PATH_BASENAME
                + "\n",
            )

            db = DirSQL(
                root=config_dir, config=os.path.join(config_dir, ".dirsql.toml")
            )
            await db.ready()
            results = await db.query(
                "SELECT basename FROM items WHERE basename = 'apple.json'"
            )
            assert len(results) == 1
            assert results[0]["basename"] == "apple.json"

            results = await db.query("SELECT * FROM items LIMIT 1")
            assert "_dirsql_file_path" not in results[0]
            assert "_dirsql_row_index" not in results[0]
